### PR TITLE
Introduce checkAccessModifier task

### LINKED ISSUE #160: No support for @Restricted annotations
Plugins built with Gradle appear to ignore `@Restricted` annotations from https://github.com/kohsuke/access-modifier/

This is dangerous as it invites the use of APIs not intended to be used outside core (or plugins) _at all_, leading to binary compatibility issues when they're changed.

### PR DESCRIPTION
**Pull request by sghill**
## Implementation

This new task delegates to `org.kohsuke.accmod.impl.Checker`, the same class relied on by the Maven plugin.

* This task runs after classes have been compiled and before the check task.
* It supports the `useBeta=true` property as gradle project property with a prefix: `checkAccessModifier.useBeta=true`. All gradle properties that start with `checkAccessModifier.` will be passed through without the prefix, so custom Restrictions are supported.
* Maven logs are bridged to the gradle logging system.

### Error Messages
Error messages are aggregated by `@Restricted` class/method, and sorted by the classes with the most call sites. A failed build will also link to the Jenkins docs
```
hudson/plugins/mercurial/MercurialChangeSet.setMsg(Ljava/lang/String;)V must not be used
    but was used on 3 lines:
        - org.example.Goodbye:38
        - org.example.Hello:11
        - org.example.Hello:23

hudson/plugins/mercurial/MercurialChangeSet.setUser(Ljava/lang/String;)V must not be used
    but was used on 1 line:
        - org.example.Goodbye:100

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':checkAccessModifier'.
> Restricted APIs were detected - see https://tiny.cc/jenkins-restricted
```

### Dependency Graph Changes
The original task implementation added several dependencies to the buildscript classpath, but this has been updated to use the Worker API with an isolated classloader.

The classloader is created from a dedicated configuration. This has the following benefits:

- It will be resolved with project repositories, buildscript repositories do not need to be updated
- Dependencies do not need to be excluded
- Other buildscript dependencies need not be conflict resolved

## Issues
There are some issues captured here in failing tests that may be issues upstream:
* Groovy code that calls `@Restricted` methods doesn't fail the Checker
* `@Restricted` on a package is allowed, but doesn't fail the Checker. Should the annotation have targets defined? (https://github.com/jenkinsci/lib-access-modifier/issues/19)

Fixes #160

### CODE REVIEW

**Comment by sghill** [comment]
Opened [JENKINS-62147](https://issues.jenkins-ci.org/browse/JENKINS-62147) regarding publishing newer versions of annotation-indexer to maven central

**Comment by timja** [comment]
Should be now unblocked assuming that maven central is not a hard requirement...

File: build.gradle.kts [hunk 1]
```diff
@@ -22,19 +25,25 @@ java {
 
 val sezpoz = "net.java.sezpoz:sezpoz:1.13"
 
+configurations.all {
+    exclude("org.jenkins-ci", "annotation-indexer")
```
Comment (timja) [inline build.gradle.kts:29]: It would be better if we could just use repo.jenkins-ci.org and not exclude this, setting up permissions and publishing there doesn't seem to be straightforward, and it's unreliably from what I hear.

Comment (sghill) [reply build.gradle.kts:29]: Thanks for the ping.

I'm hesitant to introduce a new buildscript repository requirement because I can't add that from within the plugin, so the plugin would no longer work out of the box. It may be possible to resolve these dependencies from another configuration only when the task executes. I'll take a look at what our options are.

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/AccessModifierPlugin.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                isCanBeResolved = true
+                withDependencies {
+                    add(library)
+                }
+            }
+
+            val propertyProvider = provider(PrefixedPropertiesProvider(this, CheckAccessModifierTask.PREFIX))
+            val checkAccessModifier = tasks.register<CheckAccessModifierTask>(CheckAccessModifierTask.NAME) {
+                dependsOn("classes")
+                val dirs = tasks.withType<AbstractCompile>().map { it.destinationDir }
```
Comment (tbroyer) [inline src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/AccessModifierPlugin.kt:33]: This will _realize_ all those tasks. Not necessarily a big deal but you may want to use a `provider {}` here to compute it lazily. This will also analyze test classes depending on the order Gradle decides to run the tasks because the `dirs` here doesn't have implicit task dependencies, and your explicit `dependsOn` only _selects_ part of the tasks.
If your task is run before `compileTestJava` for instance (or if that task is not run), it'll still be realized and your task will receive its `destinationDir`, which could be inexistant if the task never ran before (or could be stale).

You could use `SourceSet`s' `output.classesDirs`, which I believe will have implicit task dependencies attached to it, so `sourceSets["main"].output.classesDirs` will both implicitly declare task dependencies on `compileJava`, `compileGroovy`, and/or `compileKotlin` depending on the plugins applied to the project, **and** collect their destination directories; and without accidentally pulling in test tasks.

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        project.layout.files(compiled, dependencies)
+    }
+
+    @TaskAction
+    fun check() {
+        val q = workerExecutor.classLoaderIsolation {
+            classpath.from(accessModifierClasspath)
+        }
+        for (compilationDir in presentCompilationDirs.get()) {
+            q.submit(CheckAccess::class) {
```
Comment (tbroyer) [inline src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt:58]: An alternative to using a worker here (a that's what you were asking: https://gradle-community.slack.com/archives/CA745PZHN/p1608574892315400) would be to use an `URLClassLoader`, but because you're using a for loop here, the worker will actually parallelize work so I think it's worth it.

File: build.gradle.kts (hunk 1 above)
Comment (sghill) [reply build.gradle.kts]: Updated the PR to use the [Worker API](https://docs.gradle.org/6.7.1/userguide/custom_tasks.html#worker_api) with classloader isolation and this should work. `org.jenkins-ci:annotation-indexer` is no longer excluded.

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ * @see org.kohsuke.accmod.impl.EnforcerMojo
+ */
+open class CheckAccessModifierTask @Inject constructor(private val workerExecutor: WorkerExecutor) : DefaultTask() {
+    companion object {
+        const val NAME = "checkAccessModifier"
+        const val PREFIX = "$NAME."
+    }
+
+    @Classpath
+    val accessModifierClasspath: Property<Configuration> = project.objects.property()
```
[the remark this answers is not in the document]
Comment (sghill) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt]: This has been updated

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Internal
+    val presentCompilationDirs: Provider<List<File>> = compilationDirs.map { it.filter { f -> f.exists() } }
+
+    @Internal
+    val classpathToScan: Provider<FileCollection> = compileClasspath.map {
+        val dependencies = it.resolvedConfiguration.resolvedArtifacts.map { a -> a.file }
+        val compiled = presentCompilationDirs.get()
+        project.layout.files(compiled, dependencies)
+    }
```
[the remark this answers is not in the document]
Comment (sghill) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt]: It felt a little overcomplicated as well. I updated to these suggestions and things are looking much better.

It was unclear to me at first if I was meant to use only `ConfigurableFileCollection` or `Property<ConfigurableFileCollection>` (even though that seemed redundant). Then I saw this [in the Lazy Properties docs](https://docs.gradle.org/6.7.1/userguide/lazy_configuration.html#working_with_files_in_lazy_properties):

> All of these [collection types for File-like objects] are also considered lazy types.

I ended up instantiating it with `project.objects.fileCollection()` after seeing this note in the [JavaDoc](https://docs.gradle.org/6.7.1/javadoc/org/gradle/api/file/ConfigurableFileCollection.html):

> You can obtain an instance of ConfigurableFileCollection by calling Project.files(Object...) or ObjectFactory.fileCollection().

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessParameters.kt [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.provider.MapProperty
+import org.gradle.workers.WorkParameters
+
+interface CheckAccessParameters : WorkParameters {
+    fun getCompilationDir(): DirectoryProperty
+    fun getScannableClasspath(): ConfigurableFileCollection
+    fun getAccessModifierProperties(): MapProperty<String, Any>
```
[the remark this answers is not in the document]
Comment (sghill) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessParameters.kt]: Switched to this. I didn't realize it was supported. I see that it is in the [Custom Tasks docs](https://docs.gradle.org/6.7.1/userguide/custom_tasks.html#using-the-worker-api). The first google result for "gradle worker api" led me to [here](https://docs.gradle.org/nightly/userguide/worker_api.html#converting_to_the_worker_api), which has a java example I was mapping from.

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/AccessModifierPlugin.kt (hunk 2 above)
Comment (sghill) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/AccessModifierPlugin.kt]: Really appreciate this input! Great point about the test outputs. I've changed to it. This style also eliminates the need for checking if the directory exists.

I did have to do something a little different to get this to work:
```kotlin
project.extensions.getByType<SourceSetContainer>().getByName("main").output.classesDirs
```

I think this is the spirit of what you mentioned with this, but I believe I cannot resolve that `sourceSets` from the target project within a plugin:
```
sourceSets["main"].output.classesDirs
```

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt (hunk 3 above)
Comment (sghill) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt:41]: I think I'll stick with the worker api, but just to make sure I'm understanding the alternative is:

1. Store the current classloader in a variable
2. Update the current classloader to a new `URLClassLoader`
3. Run the task
4. Swap back to the original classloader

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/AccessModifierPlugin.kt (hunk 2 above)
Comment (tbroyer) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/AccessModifierPlugin.kt]: I think you could also write something like

```kotlin
project.the<SourceSetContainer>()["main"].output.classes
```

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt (hunk 3 above)
Comment (tbroyer) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt:41]: I think it'd be more complicated: you'd have to have the task code (that calls Kohsuke's lib) in the same class loader as the lib (not even a parent classloader, possibly a child) which generally involves writing your own classloader (I've done that a couple time, it's always a PITA; Maven's ClassWorlds make it easier; Gradle has something equivalent but it's internal).
Either that or exclusively using reflection (or if you switch the current classloader, then afaict only use classes from the lib or the jdk until you switch back).

In other words: stick to the worker API 😉

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (tbroyer) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt]: Fwiw, I've switched to using managed properties: declare the class abstract, declare the `val` abstract as well and remove initialization:

```kotlin
@CompileClasspath
abstract val compileClasspath: ConfigurableFileCollection
```

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt (hunk 3 above)
Comment (sghill) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt:41]: Thanks for the explanation

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (sghill) [reply src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt]: Thanks, I had not heard of [managed properties](https://docs.gradle.org/6.7.1/userguide/custom_gradle_types.html#managed_properties) before. Is there an advantage beyond not needing to initialize?
